File: python/examples/session_lifecycle.py
#!/usr/bin/env python3
# Copyright (c) 2024-2026 Elias Bachaalany
# SPDX-License-Identifier: MPL-2.0
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# session_lifecycle: Session management, mutations, revisions, discard, and save.
#
# Usage: python session_lifecycle.py [host_url]
#
# Demonstrates status checks, capabilities, revision tracking,
# function renaming, discard (undo), and save.

import sys

import libghidra as ghidra


def main() -> None:
    url = sys.argv[1] if len(sys.argv) >= 2 else "http://127.0.0.1:18080"
    client = ghidra.connect(url)

    # 1. Check host status
    print("--- Host status ---")
    try:
        status = client.get_status()
    except ghidra.GhidraError as e:
        print(f"Cannot reach host at {url}: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"  Service:  {status.service_name} v{status.service_version}")
    print(f"  Mode:     {status.host_mode}")
    print(f"  Revision: {status.program_revision}")
    print(f"  OK:       {status.ok}")
    if status.warnings:
        for w in status.warnings:
            print(f"  Warning:  {w}")

    # 2. List capabilities
    print("\n--- Capabilities ---")
    try:
        caps = client.get_capabilities()
        for cap in caps:
            note = f" ({cap.note})" if cap.note else ""
            print(f"  {cap.id}: {cap.status}{note}")
    except ghidra.GhidraError as e:
        print(f"GetCapabilities failed: {e}", file=sys.stderr)

    # 3. Get current revision
    print("\n--- Current revision ---")
    try:
        rev_resp = client.get_revision()
        rev_before = rev_resp.revision
        print(f"  Revision: {rev_before}")
    except ghidra.GhidraError as e:
        print(f"GetRevision failed: {e}", file=sys.stderr)
        sys.exit(1)

    # 4. Find a function to rename
    try:
        funcs = client.list_functions(limit=10)
    except ghidra.GhidraError as e:
        print(f"ListFunctions failed: {e}", file=sys.stderr)
        sys.exit(1)

    if not funcs.functions:
        print("No functions available.", file=sys.stderr)
        sys.exit(1)

    target = funcs.functions[0]
    addr = target.entry_address
    original_name = target.name

    # 5. Rename the function to create a mutation
    temp_name = "lifecycle_test_renamed"
    print(f"\n--- Renaming 0x{addr:x} '{original_name}' -> '{temp_name}' ---")
    try:
        resp = client.rename_function(addr, temp_name)
        print(f"  Renamed: {resp.renamed}, new name: {resp.name}")
    except ghidra.GhidraError as e:
        print(f"RenameFunction failed: {e}", file=sys.stderr)
        sys.exit(1)

    # 6. Check revision after mutation
    print("\n--- Revision after mutation ---")
    try:
        rev_resp = client.get_revision()
        rev_after = rev_resp.revision
        print(f"  Revision before: {rev_before}")
        print(f"  Revision after:  {rev_after}")
        print(f"  Changed: {rev_after != rev_before}")
    except ghidra.GhidraError as e:
        print(f"GetRevision failed: {e}", file=sys.stderr)

    # 7. Discard changes (undo the rename)
    print("\n--- Discarding changes ---")
    try:
        resp = client.discard_program()
        print(f"  Discarded: {resp.discarded}")
    except ghidra.GhidraError as e:
        print(f"DiscardProgram failed: {e}", file=sys.stderr)
        sys.exit(1)

    # 8. Verify the function name is restored
    print("\n--- Verifying name restored ---")
    try:
        funcs = client.list_functions(limit=10)
        restored = None
        for f in funcs.functions:
            if f.entry_address == addr:
                restored = f
                break
        if restored:
            print(f"  Address:  0x{addr:x}")
            print(f"  Name:     {restored.name}")
            print(f"  Restored: {restored.name == original_name}")
        else:
            print(f"  Function at 0x{addr:x} not found after discard.")
    except ghidra.GhidraError as e:
        print(f"ListFunctions failed: {e}", file=sys.stderr)

    # 9. Save program (commits current state to disk)
    print("\n--- Saving program ---")
    try:
        resp = client.save_program()
        print(f"  Saved: {resp.saved}")
    except ghidra.GhidraError as e:
        print(f"SaveProgram failed: {e}", file=sys.stderr)

    print("\nDone.")


if __name__ == "__main__":
    main()

File: rust/examples/session_lifecycle.rs
// Copyright (c) 2024-2026 Elias Bachaalany
// SPDX-License-Identifier: MPL-2.0
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// session_lifecycle: Check status, capabilities, and revision; mutate; observe; discard.
//
// Usage: session_lifecycle [host_url]

use libghidra as ghidra;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let url = args.get(1).map_or("http://127.0.0.1:18080", |s| s.as_str());
    let client = ghidra::connect(url);

    // 1. Check host status
    let status = client.get_status().unwrap_or_else(|e| {
        eprintln!("get_status failed: {e}");
        std::process::exit(1);
    });
    println!("Host status:");
    println!(
        "  service:  {} v{}",
        status.service_name, status.service_version
    );
    println!("  mode:     {}", status.host_mode);
    println!("  ok:       {}", status.ok);
    println!("  revision: {}", status.program_revision);
    if !status.warnings.is_empty() {
        println!("  warnings: {:?}", status.warnings);
    }

    // 2. Get capabilities
    let caps = client.get_capabilities().unwrap_or_else(|e| {
        eprintln!("get_capabilities failed: {e}");
        std::process::exit(1);
    });
    println!("\nCapabilities ({} total):", caps.len());
    for c in &caps {
        let note = if c.note.is_empty() {
            String::new()
        } else {
            format!(" -- {}", c.note)
        };
        println!("  [{}] {}{}", c.status, c.id, note);
    }

    // 3. Get current revision
    let rev_before = client.get_revision().unwrap_or_else(|e| {
        eprintln!("get_revision failed: {e}");
        std::process::exit(1);
    });
    println!("\nRevision before mutation: {}", rev_before.revision);

    // 4. Make a mutation: rename the first function
    let funcs = client
        .list_functions(0, u64::MAX, 1, 0)
        .unwrap_or_else(|e| {
            eprintln!("list_functions failed: {e}");
            std::process::exit(1);
        });
    if funcs.functions.is_empty() {
        eprintln!("No functions found. Is a program open?");
        std::process::exit(1);
    }
    let target = &funcs.functions[0];
    let original_name = target.name.clone();
    let temp_name = format!("{}_lifecycle_test", original_name);
    println!(
        "\nRenaming 0x{:x}: '{}' -> '{}'",
        target.entry_address, original_name, temp_name
    );

    let rename_resp = client
        .rename_function(target.entry_address, &temp_name)
        .unwrap_or_else(|e| {
            eprintln!("rename_function failed: {e}");
            std::process::exit(1);
        });
    println!(
        "Renamed: '{}' (renamed={})",
        rename_resp.name, rename_resp.renamed
    );

    // 5. Observe revision change
    let rev_after = client.get_revision().unwrap_or_else(|e| {
        eprintln!("get_revision (after) failed: {e}");
        std::process::exit(1);
    });
    println!("\nRevision after mutation: {}", rev_after.revision);
    if rev_after.revision > rev_before.revision {
        println!(
            "Revision advanced by {}",
            rev_after.revision - rev_before.revision
        );
    } else {
        println!("Warning: revision did not advance");
    }

    // 6. Discard all changes (restores original state)
    let discard_resp = client.discard_program().unwrap_or_else(|e| {
        eprintln!("discard_program failed: {e}");
        std::process::exit(1);
    });
    println!("\nDiscarded changes (discarded={})", discard_resp.discarded);

    // 7. Verify the function name was restored
    let verify = client
        .list_functions(0, u64::MAX, 1, 0)
        .unwrap_or_else(|e| {
            eprintln!("list_functions (verify) failed: {e}");
            std::process::exit(1);
        });
    if let Some(f) = verify.functions.first() {
        println!(
            "Verified: function at 0x{:x} is now '{}'",
            f.entry_address, f.name
        );
        if f.name == original_name {
            println!("Discard successful: name restored to original.");
        } else {
            println!(
                "Unexpected: name is '{}', expected '{}'",
                f.name, original_name
            );
        }
    }

    // 8. Final revision
    let rev_final = client.get_revision().unwrap_or_else(|e| {
        eprintln!("get_revision (final) failed: {e}");
        std::process::exit(1);
    });
    println!("\nFinal revision: {}", rev_final.revision);
}
